src/solc_0.7/GameToken/GameToken.sol

```diff
@@ -0,0 +1,558 @@
+        } else {
+            idVersion = version;
+            gameId = _generateGameId(from, subId, idVersion);
+            storageId = _storageId(gameId);
+        }
+
+        _owners[storageId] = (uint256(idVersion) << 200) + uint256(to);
+
+        emit Transfer(address(0), to, gameId);
+        return (gameId, storageId);
+    }
+
+    /// @dev Allow token owner to set game editors.
+    /// @param gameCreator The address of a GAME creator,
+    /// @param editor The address of the editor to set.
+    /// @param isEditor Add or remove the ability to edit.
+    function _setGameEditor(
+        address gameCreator,
+        address editor,
+        bool isEditor
+    ) internal {
+        emit GameEditorSet(gameCreator, editor, isEditor);
+        _gameEditors[gameCreator][editor] = isEditor;
+    }
+
+    /// @dev Bumps the version number of a game token, buring the previous
+    /// version and minting a new one.
+    /// @param from The address of the GAME token owner.
+    /// @param gameId The Game token to bump the version of.
+    /// @return The new gameId.
+    function _bumpGameVersion(address from, uint256 gameId) internal returns (uint256) {
+        address originalCreator = address(gameId / CREATOR_OFFSET_MULTIPLIER);
+        uint64 subId = uint64(gameId / SUBID_MULTIPLIER);
+        uint32 version = uint32(gameId);
+        version++;
+        address owner = _ownerOf(gameId);
+        _burn(from, owner, gameId);
+        (uint256 newId, ) = _mintGame(originalCreator, owner, subId, version, false);
+        address newOwner = _ownerOf(newId);
+        assert(owner == newOwner);
+        return newId;
+    }
+
+    /// @dev Get the baseId (full id without the version number) from the full tokenId.
+    /// @param id The full tokenId for the GAME token.
+    /// @return The baseId.
+    function _storageId(uint256 id) internal pure override returns (uint256) {
+        return uint256(id & BASEID_MASK);
+    }
+
+    /// @dev Create a new gameId and associate it with an owner.
+    /// This is a packed id, consisting of 3 parts:
+    /// the creator's address, a uint64 subId and a uint32 version number.
+    /// @param creator The address of the Game creator.
+    /// @param subId The id to use when generating the new GameId.
+    function _generateGameId(
+        address creator,
+        uint64 subId,
+        uint32 version
+    ) internal pure returns (uint256) {
+        return uint256(creator) * CREATOR_OFFSET_MULTIPLIER + uint64(subId) * SUBID_MULTIPLIER + uint32(version);
+    }
+
+    /// @dev Get the a full URI string for a given hash + gameId.
+    /// @param hash The 32 byte IPFS hash.
+    /// @param id The token Id for the GAME.
+    /// @return The URI string.
+    function toFullURI(bytes32 hash, uint256 id) internal pure returns (string memory) {
+        return string(abi.encodePacked("ipfs://bafybei", hash2base32(hash), "/", id.toString(), ".json"));
```

Comment:
I think the id is unecessary.
we could have ipfs://..../game.json

At the same time have the <id>.json can be suseful when you download such file, but that is not really a common use

Reply:
@wighawag So a complete URI string example could look like this:

'ipfs://bafybeibjjuirh6mb3gq36smqasbpz5vxf2ftxs2ezydxiauwkezw5abaqy/game.json'

Reply:
Or do you prefer that we drop the 'game.json' part from the string?

Reply:
game.json is good as this allow ipfs gateway to use the correct mime type for .json